File: apps/web/src/lib/watch-status.test.ts
import { describe, expect, it } from "vitest"
import type { WatchProgress } from "./api"
import {
  completionEpisodeIds,
  episodeProgressPercent,
  episodeWatchState,
  posterWatchState,
  resumePositionLabel,
  seasonWatchVideos,
  seriesWatchVideos,
} from "./watch-status"

function progress(values: Partial<WatchProgress> & Pick<WatchProgress, "videoId">): WatchProgress {
  return {
    videoId: values.videoId,
    mediaType: values.mediaType ?? "series",
    mediaId: values.mediaId ?? "show",
    name: "Show",
    positionMs: values.positionMs ?? 0,
    durationMs: values.durationMs ?? 0,
    watched: values.watched ?? false,
    updatedAt: "2026-01-01T00:00:00Z",
  }
}

describe("poster watch state", () => {
  it("treats movies as a binary watched state", () => {
    const movie = { id: "movie", type: "movie" }
    expect(posterWatchState([], movie)).toBe("unwatched")
    expect(
      posterWatchState(
        [progress({ videoId: "movie", mediaId: "movie", mediaType: "movie", watched: true })],
        movie,
      ),
    ).toBe("complete")
  })

  it("distinguishes partial and automatically complete series state", () => {
    const series = { id: "show", type: "series" }
    expect(posterWatchState([progress({ videoId: "show:1:1", positionMs: 30_000 })], series))
      .toBe("partial")
    expect(
      posterWatchState(
        [
          progress({ videoId: "show:1:1", watched: true }),
          progress({ videoId: "show:1:2", watched: true }),
        ],
        series,
        ["show:1:1", "show:1:2"],
      ),
    ).toBe("complete")
  })

  it("does not require specials, unavailable, or future episodes for completion", () => {
    expect(
      completionEpisodeIds(
        [
          { id: "regular", season: 1, episode: 1, released: "2026-01-01" },
          { id: "special", season: 0, episode: 1 },
          { id: "unavailable", season: 1, episode: 2, available: false },
          { id: "future", season: 1, episode: 3, released: "2027-01-01" },
        ],
        Date.parse("2026-06-01"),
      ),
    ).toEqual(["regular"])
  })
})

describe("episode watch state", () => {
  it("distinguishes not started, in progress, and watched", () => {
    expect(episodeWatchState()).toBe("not-started")
    expect(episodeWatchState(progress({ videoId: "partial", positionMs: 25, durationMs: 100 })))
      .toBe("in-progress")
    expect(episodeProgressPercent(progress({ videoId: "partial", positionMs: 125, durationMs: 100 })))
      .toBe(100)
    expect(episodeWatchState(progress({ videoId: "complete", watched: true }))).toBe("watched")
    expect(episodeProgressPercent(progress({ videoId: "complete", watched: true, positionMs: 1, durationMs: 2 })))
      .toBe(0)
  })

  it("formats an unfinished position for resume prompts", () => {
    expect(resumePositionLabel(progress({ videoId: "short", positionMs: 6_900 }))).toBe("0:06")
    expect(resumePositionLabel(progress({ videoId: "long", positionMs: 65_000 }))).toBe("1:05")
    expect(resumePositionLabel(progress({ videoId: "hour-long", positionMs: 3_908_000 }))).toBe("1:05:08")
    expect(resumePositionLabel()).toBeUndefined()
    expect(resumePositionLabel(progress({ videoId: "watched", positionMs: 6_000, watched: true })))
      .toBeUndefined()
  })

  it("selects every released episode in one season", () => {
    const videos = [
      { id: "s1e1", season: 1, episode: 1, released: "2026-01-01" },
      { id: "s1e2", season: 1, episode: 2, released: "2026-01-01" },
      { id: "s1e3", season: 1, episode: 3, released: "2027-01-01" },
      { id: "s2e1", season: 2, episode: 1, released: "2026-01-01" },
      { id: "unavailable", season: 1, episode: 4, available: false },
    ]
    expect(seasonWatchVideos(videos, 1, Date.parse("2026-06-01")))
      .toEqual([videos[0], videos[1]])
  })

  it("uses regular episodes for series actions while preserving specials-only titles", () => {
    const regular = [
      { id: "special", season: 0, episode: 1 },
      { id: "episode", season: 1, episode: 1 },
    ]
    expect(seriesWatchVideos(regular).map((video) => video.id)).toEqual(["episode"])
    expect(completionEpisodeIds([{ id: "special", season: 0, episode: 1 }])).toEqual(["special"])
  })
})

File: apps/mobile/composeApp/src/commonTest/kotlin/media/conduit/mobile/WatchStatusTest.kt
package media.conduit.mobile

import kotlin.test.Test
import kotlin.test.assertEquals
import media.conduit.mobile.account.CatalogItem
import media.conduit.mobile.account.ProgressSummary
import media.conduit.mobile.account.VideoItem

class WatchStatusTest {
    @Test
    fun moviesUseBinaryCompletion() {
        val movie = CatalogItem("movie", "movie", "Movie")
        assertEquals(PosterWatchState.Unwatched, posterWatchState(emptyList(), movie))
        assertEquals(PosterWatchState.Complete, posterWatchState(listOf(progress("movie", "movie", watched = true, type = "movie")), movie))
    }

    @Test
    fun seriesDistinguishPartialAndComplete() {
        val series = CatalogItem("show", "series", "Show")
        assertEquals(PosterWatchState.Partial, posterWatchState(listOf(progress("s1e1", "show", position = 30_000)), series, listOf("s1e1", "s1e2")))
        assertEquals(PosterWatchState.Complete, posterWatchState(listOf(progress("s1e1", "show", watched = true), progress("s1e2", "show", watched = true)), series, listOf("s1e1", "s1e2")))
    }

    @Test
    fun completionExcludesSpecialsAndFutureEpisodes() {
        assertEquals(
            listOf("regular"),
            completionEpisodeIds(
                listOf(
                    VideoItem("regular", season = 1, episode = 1, released = "2026-01-01"),
                    VideoItem("special", season = 0, episode = 1),
                    VideoItem("unavailable", season = 1, episode = 2, available = false),
                    VideoItem("future", season = 1, episode = 2, released = "2027-01-01"),
                ),
                today = "2026-06-01",
            ),
        )
    }

    @Test
    fun episodeStateAndProgressUseExplicitWatchedFlag() {
        assertEquals(EpisodeWatchState.NotStarted, episodeWatchState(null))
        val partial = progress("partial", "show", position = 25).copy(durationMs = 100_000)
        assertEquals(EpisodeWatchState.InProgress, episodeWatchState(partial))
        assertEquals(.25f, episodeProgressFraction(partial))
        assertEquals(EpisodeWatchState.Watched, episodeWatchState(partial.copy(watched = true)))
        assertEquals(0f, episodeProgressFraction(partial.copy(watched = true)))
    }

    @Test
    fun restOfSeasonStartsAtCurrentEpisodeAndSkipsUnavailableEpisodes() {
        val videos = listOf(
            VideoItem("s1e1", season = 1, episode = 1, released = "2026-01-01"),
            VideoItem("s1e2", season = 1, episode = 2, released = "2026-01-01"),
            VideoItem("future", season = 1, episode = 3, released = "2027-01-01"),
            VideoItem("s2e1", season = 2, episode = 1, released = "2026-01-01"),
        )

        assertEquals(
            listOf("s1e2"),
            restOfSeasonWatchVideos(videos, 1, "s1e2", today = "2026-06-01").map(VideoItem::id),
        )
    }

    @Test
    fun detailsResumeUnfinishedMovies() {
        val movie = CatalogItem("movie", "movie", "Movie")
        val unfinished = progress("movie", "movie", position = 30_000, type = "movie")

        assertEquals(unfinished, latestUnfinishedProgress(listOf(unfinished), movie))
        assertEquals("Resume", detailsPlayLabel(movie, unfinished, null))
        assertEquals("Play", detailsPlayLabel(movie, unfinished.copy(watched = true), null))
    }

    @Test
    fun detailsResumeSeriesWithEpisodeNumber() {
        val series = CatalogItem("show", "series", "Show")
        val unfinished = progress("s1e2", "show", position = 30_000)
        val episode = VideoItem("s1e2", season = 1, episode = 2)

        assertEquals(unfinished, latestUnfinishedProgress(listOf(unfinished), series))
        assertEquals("Resume S1E2", detailsPlayLabel(series, unfinished, episode))
        assertEquals("Play", detailsPlayLabel(series, unfinished, VideoItem("s1e3", season = 1, episode = 3)))
    }

    @Test
    fun zeroProgressDoesNotOfferResume() {
        val movie = CatalogItem("movie", "movie", "Movie")
        val notStarted = progress("movie", "movie", position = 0, type = "movie")

        assertEquals(null, latestUnfinishedProgress(listOf(notStarted), movie))
        assertEquals("Play", detailsPlayLabel(movie, null, null))
    }

    @Test
    fun explicitHistoryEpisodeWinsAndMissingEpisodeFallsBackToLatestUnfinished() {
        val series = CatalogItem("show", "series", "Show")
        val unfinished = progress("s1e2", "show", position = 30_000)

        assertEquals("s1e3", effectiveResumeVideoId("s1e3", listOf(unfinished), series))
        assertEquals("s1e2", effectiveResumeVideoId(null, listOf(unfinished), series))
    }

    private fun progress(videoId: String, mediaId: String, watched: Boolean = false, position: Long = 0, type: String = "series") =
        ProgressSummary(videoId, type, mediaId, "Title", positionMs = position, durationMs = 100_000, watched = watched, updatedAt = "2026-01-01")
}
